File: odroid/Share/src/bert/share/message/MessageBottle.java
/**
 * Copyright 2019 Charles Coughlin. All rights reserved.
 * MIT License
 */
package bert.share.message;

import java.io.IOException;
import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

import com.fasterxml.jackson.core.JsonParseException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * This class represents the contents of requests and responses that are sent
 * across sockets between parts of the robot. it becomes serialized into JSON.
 * 
 * There is no intrinsic difference between requests and
 * responses. We leave it to context to determine which is which.
 * 
 * Leave members public to be accessible via Reflection. We use
 * fetch/assign instead of get/set for the shortcut methods that
 * access the properties to avoid confusion by the JSON mapper.
 */
public class MessageBottle implements Serializable {
	private static final long serialVersionUID = 4356286171135500644L;
	private static final String CLSS = "MessageBottle";
	protected static final Logger LOGGER = Logger.getLogger(CLSS);
	public Map<String,String> properties;  // Multiple properties for a single motor
	public Map<String,String> jointValues; // A single property for multiple motors 
	
	public MessageBottle() {
		this.properties = new HashMap<>();
		this.jointValues = new HashMap<>();
	}
	
	public String getJointValue(String joint,String defaultValue) {
		String value = this.jointValues.get(joint);
		if( value==null ) value = defaultValue;
		return value;
	}
	
	public void setJointValue(String joint,String value) {		
		this.jointValues.put(joint,value);
	}
	
	public Map<String,String>  getProperties() {return this.properties;}
	public Map<String,String>  getJointValues() {return this.jointValues;}
	
	public String getProperty(String key,String defaultValue) {
		String value = this.properties.get(key);
		if( value==null ) value = defaultValue;
		return value;
	}
	
	public void setProperty(String key,String value) {
		this.properties.put(key, value);
	}

	/**
	 * Set the value map all at once. The values are expected to correspond to
	 * the stated property.
	 * @param values the new value map.
	 */
	public void setJointValues(Map<String,String> values) { this.jointValues = values; }
	
	
	/**
	 * Use this method to clear the message properties. This conveniently
	 * allows re-use of the message object when issuing a new command.
	 */
	public void clear() {
		this.properties.clear();
	}
	/**
	 * This is a convenience method retrieve an error message property.
	 * The message should be suitable for direct display to the user.
	 * 
	 * @return an error message. If there is no error message, return null.
	 */
	public String fetchError() {
		return getProperty(BottleConstants.ERROR,null);
	}
	
	/**
	 * This is a convenience method to set a string suitable for audio
	 * that indicates an error. If an error is present no further
	 * processing is valid.
	 * 
	 * @param msg a message suitable to be played for the user.
	 */
	public void assignError(String msg) {
		setProperty(BottleConstants.ERROR, msg);
	}
	
	/**
	 * For a message that is a request, the request type should be set using
	 * the setter supplied in this class.
	 * 
	 * @return the RequestType. If not set, return NONE.
	 */
	public RequestType fetchRequestType() {
		RequestType type = RequestType.NONE;
		String prop = getProperty(BottleConstants.TYPE,null);
		if( prop!=null) {
			type = RequestType.valueOf(prop);
		}
		return type;
	}
	
	/**
	 * For a message that is a request, use this method to set its type. This
	 * is our way of enforcing a fixed vocabulary.
	 * 
	 * @param type the type of request.
	 */
	public void assignRequestType(RequestType type) {
		setProperty(BottleConstants.TYPE, type.name());
	}

	
	/**
	 * Convenience method to retrieve the ControllerType of the message source.
	 * 
	 * @return a source name. If there is no identified source, return null.
	 */
	public String fetchSource() {
		return getProperty(BottleConstants.SOURCE,null);
	}
	
	/**
	 * Convenience method to set a string naming the message creator. Use
	 * the controller type for this.
	 * 
	 * @param source the name of the message creator.
	 */
	public void assignSource(String source) {
		setProperty(BottleConstants.SOURCE, source);
	}
	
	// =================================== JSON ======================================
	public static MessageBottle fromJSON(String json) {
		MessageBottle bottle = null;
		ObjectMapper mapper = new ObjectMapper();
		try {
			bottle = mapper.readValue(json,MessageBottle.class);
			
		}
		catch(JsonParseException jpe) {
			LOGGER.severe(String.format("%s.fromJSON: Parse exception (%s) from %s",CLSS,jpe.getLocalizedMessage(),json));
		}
		catch(JsonMappingException jme) {
			LOGGER.severe(String.format("%s.fromJSON: Mapping exception (%s) from %s",CLSS,jme.getLocalizedMessage(),json));
		}
		catch(IOException ioe) {
			LOGGER.severe(String.format("%s.fromJSON: IO exception (%s)",CLSS,ioe.getLocalizedMessage()));
		}
		
		return bottle;
	}
	
	public String toJSON() {
		ObjectMapper mapper = new ObjectMapper();
		mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
		String json="";
		try {
			json = mapper.writeValueAsString(this);
			LOGGER.info(String.format("%s.toJSON: [%s]",CLSS,json));
		}
		catch(Exception ex) {
			LOGGER.severe(String.format("%s.toJSON: Exception (%s)",CLSS,ex.getLocalizedMessage()));
		}
		return json;
	}
}

File: Bert/common/src/main/kotlin/chuckcoughlin/bert/common/message/MessageBottle.kt
/**
 * Copyright 2022-2024 Charles Coughlin. All rights reserved.
 * MIT License
 */
package chuckcoughlin.bert.common.message

import chuckcoughlin.bert.common.controller.ControllerType
import chuckcoughlin.bert.common.model.*
import com.fasterxml.jackson.core.JsonParseException
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.JsonMappingException
import com.fasterxml.jackson.module.kotlin.*
import java.io.IOException
import java.io.Serializable
import java.util.logging.Logger

/**
 * This class is a container for requests and responses that are sent
 * across channels between parts of the robot. Semantics are entirely
 * determined by the message type.
 *
 * There is no intrinsic difference between requests and
 * responses. We leave it to context to determine which is which.
 * There may be multiple affected joints and multiple properties.
 * The collection of properties determines whether or not the
 * message makes practical sense.
 *
 * The message affects either a single or multiple joints, and may change
 * multiple dynamic properties at a time.
 *
 *  Sets bottle type. The rest are settable
*/
data class MessageBottle (var type:RequestType) : Cloneable,Serializable {
    var extremity: Extremity // Message applies to this extremity
    var command : CommandType
    var error : String       // Error message if not blank
    var handler: ControllerType // The subsystem to handle this message
    var id: Long             // Unique identifier (used for serial processing)
    var joint : Joint        // Request applies to this joint (if one)
    var jtype: JsonType      // Type when message is JSON
    var limb: Limb           // Message applies to this limb
    var metric: MetricType
    var arg: String          // Use whenever the command requires a text argument
    var jointDefinitionProperty: JointDefinitionProperty   // Possible subject of the original request
    var jointDynamicProperty: JointDynamicProperty         // Possible subject of the original request
    var source: ControllerType       // Origin of the request
    var text : String        // Pronounceable text of a response
    var value: Double
    var control : ExecutionControl  // Parameters dealing with execution of the message

    /**
     * Set the number of millisecs that the motion commanded by this
     * request is expected to last. Often this represents the required
     * time buffer between subsequent commands, especially those affecting
     * the same sub-chain.
     * @param period ~ msecs
     */
    var duration: Long = 0 // ~msecs

    override public fun clone(): MessageBottle {
        val copy = MessageBottle(type)
        copy.extremity   = extremity
        copy.command     = command
        copy.error       = error
        copy.handler     = handler
        copy.joint       = joint
        copy.jtype       = jtype
        copy.limb        = limb
        copy.metric      = metric
        copy.arg         = arg
        copy.jointDefinitionProperty = jointDefinitionProperty
        copy.jointDynamicProperty    = jointDynamicProperty
        copy.source      = source
        copy.text        = text
        copy.control     = control.clone()
        return copy
    }

    // =================================== JSON ======================================
    fun toJSON(): String {
        val mapper = jacksonObjectMapper()
        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        var json = ""
        try {
            json = mapper.writeValueAsString(this)
            LOGGER.info(String.format("%s.toJSON: [%s]", CLSS, json))
        }
        catch (ex: Exception) {
            LOGGER.severe(String.format("%s.toJSON: Exception (%s)", CLSS, ex.localizedMessage))
        }
        return json
    }

    companion object {
        private const val serialVersionUID = 4356286171135500644L
        protected const val CLSS = "MessageBottle"
        protected val LOGGER = Logger.getLogger(CLSS)

        // =================================== JSON ======================================
        fun fromJSON(json: String): MessageBottle? {
            var bottle: MessageBottle? = null
            val mapper = jacksonObjectMapper()
            try {
                bottle = mapper.readValue(json)
            }
            catch (jpe: JsonParseException) {
                LOGGER.severe(String.format("%s.fromJSON: Parse exception (%s) from %s",
                        CLSS, jpe.getLocalizedMessage(), json
                    )
                )
            }
            catch (jme: JsonMappingException) {
                LOGGER.severe(String.format("%s.fromJSON: Mapping exception (%s) from %s",
                        CLSS, jme.getLocalizedMessage(), json
                    )
                )
            }
            catch (ioe: IOException) {
                LOGGER.severe(String.format("%s.fromJSON: IO exception (%s)", CLSS, ioe.getLocalizedMessage()))
            }
            return bottle
        }
    }

    /**
     * Set initiial values for all message parameters
     */
    init {
        extremity   = Extremity.NONE
        command     = CommandType.NONE
        error   =  BottleConstants.NO_ERROR   // No error
        handler = ControllerType.UNDEFINED
        id = 0
        joint = Joint.NONE                  // Name of relavant joint
        jtype = JsonType.UNDEFINED
        limb  = Limb.NONE
        metric = MetricType.NAME
        arg    = BottleConstants.NO_ARG
        jointDefinitionProperty = JointDefinitionProperty.NONE
        jointDynamicProperty = JointDynamicProperty.NONE
        source = ControllerType.UNDEFINED
        text  = ""   // Text is the printable response
        value = Double.NaN
        control = ExecutionControl(BottleConstants.NO_DELAY)
    }
}